### PR TITLE
Overhaul packages.yaml

### LINKED ISSUE #1115: Update common/packages.yaml
For the 1.8.0 release, I think it would be a good idea to run the check_package_config utility on a couple machines and go through the output to reconcile the versions/variants in there against what actually gets concretized. For example, in testing #1109, concretizing unified-dev led to py-h5py@3.11.0 being spec'd in spite of 3.7.0 being preferred in common/packages.yaml. In those sorts of situations, we should either require the older version if needed, or just move to the/a newer version.

Some mismatches:
 - py-h5py version
 - proj version
 - scotch `metis` variant (see https://github.com/spack/spack/issues/44676)

### PR DESCRIPTION
**Pull request by AlexanderRichert-NOAA**
### Summary

This PR updates common/packages.yaml to remove unused entries, unpin versions that don't need to be pinned, and apply hard requirements as much as possible.

### Status:
99% of the entries now use `require:`.
 ```
bacio:
  version: ['2.4.1']
```
becomes
```
bacio:
  require: '@2.4.1'
```
Then if someone wants to use a different version in their own environment (or site config), they have to override it (note the ::):
```
bacio::
  require: '@2.5.0'
```

### Testing
_Make sure to check for duplicates!_
 - Tested on Alex's PC
 - Concretizes without duplicates on Hera (required '::' for ecflow entry in site config)
 - [x] Hera/Intel/UFSWM (@AlexanderRichert-NOAA)
 - [x] Hera/Intel/build full unified env (@AlexanderRichert-NOAA)
 - [x] Hera/GCC/concretization (@AlexanderRichert-NOAA)
 - [x] Gaea C5/Intel/UFSWM (@AlexanderRichert-NOAA) - ran into an unrelated issue with mapl ^cray-mpich, captured separately in #1153
 - [x] Hercules/Intel/concretization (@AlexanderRichert-NOAA)
 - [x] Hercules/GCC/concretization (@AlexanderRichert-NOAA)
 - [x] Jet/Intel/concretization (@AlexanderRichert-NOAA)
 - [x] Jet/GCC/concretization (@AlexanderRichert-NOAA)

### Applications affected

All

### Systems affected

All

### Dependencies

none

### Issue(s) addressed

Fixes #1115

### Checklist
- [x] This PR addresses one issue/problem/enhancement, or has a very good reason for not doing so.
- [ ] These changes have been tested on the affected systems and applications.
- [ ] All dependency PRs/issues have been resolved and this PR can be merged.

### CODE REVIEW

**Comment by climbfuji** [comment]
I am wondering if it makes sense to wait until we've relaxed all versions that we don't really care about?

**Comment by AlexanderRichert-NOAA** [comment]
That's fine with me... Would it make sense to go through it in its current state and see what we actually need?

**Comment by climbfuji** [comment]
> That's fine with me... Would it make sense to go through it in its current state and see what we actually need?

Yes, absolutely - if you don't mind, please go ahead! Thanks very much.

**Comment by climbfuji** [comment]
@AlexanderRichert-NOAA You may want to hold back until we've repaired the Ubuntu CI runner

**Comment by climbfuji** [comment]
@AlexanderRichert-NOAA ubuntu ci is good again. macos ci may be shaky now, but ignore it please

File: .github/workflows/ubuntu-ci-x86_64-gnu.yaml
Comment (climbfuji) [inline .github/workflows/ubuntu-ci-x86_64-gnu.yaml]: Since you changed $PWD/envs/${ENVNAME} for macOS CI, do you also want to do it for Ubuntu CI?

File: .github/workflows/ubuntu-ci-x86_64-intel.yaml
Comment (climbfuji) [inline .github/workflows/ubuntu-ci-x86_64-intel.yaml]: and same here

File: configs/common/packages.yaml
Comment (climbfuji) [inline configs/common/packages.yaml]: This is not going to work. When we move to oneapi, we need to use an external bison since oneapi doesn't build bison correctly (conflict in the package). We'll have to accept whatever `bison@3` that is found on the system.

File: configs/common/packages.yaml
Comment (climbfuji) [inline configs/common/packages.yaml:42]: Does this work with external ecflow packages?

File: configs/common/packages.yaml
Comment (climbfuji) [reply configs/common/packages.yaml]: or maybe `::` for the external package? I'd prefer just setting `require: '@3'` if possible

File: configs/common/packages.yaml
Comment (climbfuji) [inline configs/common/packages.yaml]: nice!

File: configs/common/packages.yaml
Comment (climbfuji) [inline configs/common/packages.yaml:77]: Does this work with the vastly different versions of FMS, where the older ones  don't even know about some of these variants?

File: configs/common/packages.yaml
Comment (climbfuji) [inline configs/common/packages.yaml]: let's remove 89 and 90 if not needed

File: configs/common/packages.yaml
Comment (climbfuji) [inline configs/common/packages.yaml]: let's remove that

File: configs/common/packages.yaml
Comment (climbfuji) [inline configs/common/packages.yaml:196]: I need to test this. Chances are that this will break the neptune-dev +python environment, which currently isn't tested in CI

File: configs/common/packages.yaml
Comment (climbfuji) [inline configs/common/packages.yaml]: Shall we remove 200-202?

File: configs/common/packages.yaml
Comment (climbfuji) [inline configs/common/packages.yaml]: let's remove 260-261 if you don't mind

**Review by climbfuji — CHANGES_REQUESTED** [review]
Thanks for your hard work on this. We need to do thorough testing on this PR:
- @climbfuji needs to build the neptune standalone env with the python variant enabled and test NEPTUNE with it
- JCSDA needs to test if the updated unified/skylab environment still works with JEDI
- EMC/EPIC - UFS testing?
- GMAO - any GEOS testing?

We definitely need to run the container build tests for this PR. But first we need to update `configs/containers/specs/jedi-ci.yaml` to match what is built in the unified environment / skylab environment.

@srherbener @RatkoVasic-NOAA @mathomp4 FYI

File: configs/common/packages.yaml
Comment (mathomp4) [reply configs/common/packages.yaml]: Huh. This might be one of the reasons spack-stack GEOS is non-zero-diff. Never noticed that...

**Comment by mathomp4** [comment]
> @srherbener @RatkoVasic-NOAA @mathomp4 FYI

I'm willing to try it out...if I can learn how. 😄 I'm currently at the "I know how to load and build with spack-stack" stage. I guess I need to advance to the "test a PR" stage...

File: configs/common/packages.yaml
Comment (AlexanderRichert-NOAA) [reply configs/common/packages.yaml:42]: It needs to be fully tested, but yes-- the idea is that each site config will use `ecflow::`to override.

File: configs/common/packages.yaml
Comment (AlexanderRichert-NOAA) [reply configs/common/packages.yaml]: For now, I'm setting it to '@3'. Alternatively, we could try using the same trick as with ecflow, i.e., use `bison::` in the site configs to override the common setting.

**Comment by AlexanderRichert-NOAA** [comment]
@climbfuji this seems to be in pretty good shape. Ideally it would be good to run at least a concretization on each platform just make sure it doesn't fail and to check for duplicates.

**Comment by climbfuji** [comment]
> @climbfuji this seems to be in pretty good shape. Ideally it would be good to run at least a concretization on each platform just make sure it doesn't fail and to check for duplicates.

Thanks @AlexanderRichert-NOAA. I'll create a test environment for NEPTUNE on one of the NRL machines to run some tests; I encourage others (@RatkoVasic-NOAA @srherbener @mathomp4) to do the same with their applications.

**Comment by climbfuji** [comment]
FYI I'm having issues on Nautilus which uses mkl as blas, fftw-api and lapack provider. Something insists on netlib-lapack and doesn't want to use mkl (that's not the case for develop, so must be a new/different version of one of the packages).

**Comment by climbfuji** [comment]
> FYI I'm having issues on Nautilus which uses mkl as blas, fftw-api and lapack provider. Something insists on netlib-lapack and doesn't want to use mkl (that's not the case for develop, so must be a new/different version of one of the packages).

I resolved those issues for the unified environment on Nautilus with Intel (classic). No clean solution that works for all compilers. Now running into issues with the neptune standalone environment :-(

**Comment by srherbener** [comment]
I tried building on my Mac - I have Sonoma 14.4.1 with apple-clang%15.0.0 and I'm running into the musl build issue again. Concretize successfully completed and I see crtm, esmf and fms show up in the duplicates list:
```
MacBook-Pro-5:unified-env.mymacos steveherbener$ ../../util/show_duplicate_packages.py -d log.concretize 
z35pjfx  crtm@2.4.0.1%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  +fix~ipo  build_system=cmake  build_type=Release  generator=make  arch=darwin-sonoma-m1
p563nq7  crtm@v2.4.1-jedi%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  +fix~ipo  build_system=cmake  build_type=Release  generator=make  arch=darwin-sonoma-m1
frvnpsl  crtm@v2.4-jedi.2%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  +fix~ipo  build_system=cmake  build_type=Release  generator=make  arch=darwin-sonoma-m1
bzhk57c  esmf@8.6.1%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  ~debug~external-lapack+external-parallelio+mpi+netcdf~pnetcdf+shared~xerces  build_system=makefile  esmf_comm=auto  esmf_os=auto  esmf_pio=auto  patches=f63d405  snapshot=none  arch=darwin-sonoma-m1
omhbmud  esmf@8.7.0b04%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  ~debug~external-lapack+external-parallelio+mpi+netcdf~pnetcdf+shared~xerces  build_system=makefile  esmf_comm=auto  esmf_os=auto  esmf_pio=auto  patches=f63d405  snapshot=b04  arch=darwin-sonoma-m1
gvbdrwc  fms@release-jcsda%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  +gfs_phys+internal_file_nml~ipo~large_file+openmp+quad_precision  build_system=cmake  build_type=Release  generator=make  precision=32  arch=darwin-sonoma-m1
h5vv6ws  fms@2023.04%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  +deprecated_io+gfs_phys+internal_file_nml~ipo~large_file+openmp+pic+quad_precision~yaml  build_system=cmake  build_type=Release  constants=GFS  generator=make  precision=32,64  arch=darwin-sonoma-m1
===
Duplicates found!
```
Are these the expected duplicates?

Also I can try testing on S4 for JCSDA. @climbfuji what is the method for doing this. Do I become the jedipara user and then build in a directory off to the side, test the installation out, then remove the build? Thanks!

**Comment by climbfuji** [comment]
> I tried building on my Mac - I have Sonoma 14.4.1 with apple-clang%15.0.0 and I'm running into the musl build issue again. Concretize successfully completed and I see crtm, esmf and fms show up in the duplicates list:
> 
> ```
> MacBook-Pro-5:unified-env.mymacos steveherbener$ ../../util/show_duplicate_packages.py -d log.concretize 
> z35pjfx  crtm@2.4.0.1%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  +fix~ipo  build_system=cmake  build_type=Release  generator=make  arch=darwin-sonoma-m1
> p563nq7  crtm@v2.4.1-jedi%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  +fix~ipo  build_system=cmake  build_type=Release  generator=make  arch=darwin-sonoma-m1
> frvnpsl  crtm@v2.4-jedi.2%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  +fix~ipo  build_system=cmake  build_type=Release  generator=make  arch=darwin-sonoma-m1
> bzhk57c  esmf@8.6.1%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  ~debug~external-lapack+external-parallelio+mpi+netcdf~pnetcdf+shared~xerces  build_system=makefile  esmf_comm=auto  esmf_os=auto  esmf_pio=auto  patches=f63d405  snapshot=none  arch=darwin-sonoma-m1
> omhbmud  esmf@8.7.0b04%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  ~debug~external-lapack+external-parallelio+mpi+netcdf~pnetcdf+shared~xerces  build_system=makefile  esmf_comm=auto  esmf_os=auto  esmf_pio=auto  patches=f63d405  snapshot=b04  arch=darwin-sonoma-m1
> gvbdrwc  fms@release-jcsda%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  +gfs_phys+internal_file_nml~ipo~large_file+openmp+quad_precision  build_system=cmake  build_type=Release  generator=make  precision=32  arch=darwin-sonoma-m1
> h5vv6ws  fms@2023.04%apple-clang@15.0.0  ldflags=-Wl,-ld_classic  +deprecated_io+gfs_phys+internal_file_nml~ipo~large_file+openmp+pic+quad_precision~yaml  build_system=cmake  build_type=Release  constants=GFS  generator=make  precision=32,64  arch=darwin-sonoma-m1
> ===
> Duplicates found!
> ```
> 
> Are these the expected duplicates?
> 
> Also I can try testing on S4 for JCSDA. @climbfuji what is the method for doing this. Do I become the jedipara user and then build in a directory off to the side, test the installation out, then remove the build? Thanks!

The musl issue is related to your libiconv not being found/accepted as an external package and as the only provider for iconv if I recall correctly. Not related to this PR. For S4, why jedipara? You can test with your own user as well, or?

**Comment by climbfuji** [comment]
Also, @srherbener please note that there is a follow-up PR #1159 that you should use for testing; it will eventually be merged into this PR.

File: configs/common/packages.yaml
Comment (climbfuji) [reply configs/common/packages.yaml:196]: this seems to be ok

**Comment by climbfuji** [comment]
> Also, @srherbener please note that there is a follow-up PR #1159 that you should use for testing; it will eventually be merged into this PR.



> Also, @srherbener please note that there is a follow-up PR #1159 that you should use for testing; it will eventually be merged into this PR.

#1159 was merged into this PR, thus #1138 is the best version to test (if you started already with #1159, no worries; they are identical).

**Comment by AlexanderRichert-NOAA** [comment]
@climbfuji using `packages:all:prefer:['%oneapi']` took care of the oneapi issues, in that gcc-runtime and bison use %gcc but everything else uses %oneapi.

**Comment by climbfuji** [comment]
> @climbfuji using `packages:all:prefer:['%oneapi']` took care of the oneapi issues, in that gcc-runtime and bison use %gcc but everything else uses %oneapi.

If the compiler you want to build with is `oneapi`, I assume?

**Comment by AlexanderRichert-NOAA** [comment]
Yes

**Comment by climbfuji** [comment]
> Yes

Thanks. I'll add this to `spack stack create` when I finally get to work on the single compiler logic

File: configs/common/packages.yaml
Comment (climbfuji) [inline configs/common/packages.yaml:139]: Is that because the two_level_namespace option in mpich no longer exists?

**Review by climbfuji — APPROVED** [review]
I am happy with the PR as is. Any follow-up site updates and other tweaks (e.g. the bison%gcc build for oneAPI) can be made later. NEPTUNE builds and runs with this setup, and the timing is comparable to the current spack-stack develop environment.

**Comment by srherbener** [comment]
@AlexanderRichert-NOAA, @climbfuji I am building spack-stack on S4 now. I had the `--source` option on the install command (as written in the documentation) and rapidly used up my quota of inodes. I'm trying now without the `--source` option.

The concretize step worked just fine. Install is running.

jedi-bundle is currently broken so I'm not sure when I can test that and skylab. Perhaps it's best to call it good with the successful concretize, and see if the install works. I don't want to unnecessarily hold up this PR.

What do you think?

**Comment by climbfuji** [comment]
> @AlexanderRichert-NOAA, @climbfuji I am building spack-stack on S4 now. I had the `--source` option on the install command (as written in the documentation) and rapidly used up my quota of inodes. I'm trying now without the `--source` option.
> 
> The concretize step worked just fine. Install is running.
> 
> jedi-bundle is currently broken so I'm not sure when I can test that and skylab. Perhaps it's best to call it good with the successful concretize, and see if the install works. I don't want to unnecessarily hold up this PR.
> 
> What do you think?

We should update the instructions to not use `--source` by default. I am ok with moving ahead once your install finishes.

File: configs/common/packages.yaml
Comment (srherbener) [reply configs/common/packages.yaml:139]: Looks like two_level_namespace is always enabled on MacOS so I think we are good.

https://github.com/JCSDA/spack/blob/8ba0bbae6698adf7dce525c0cf72c29edf5fb254/var/spack/repos/builtin/packages/mpich/package.py#L552-L553

Comment (climbfuji) [reply configs/common/packages.yaml:139]: Perfect, thanks. Not that it matters. We can't use mpich on macOS since it doesn't support mpi_f08.

**Comment by srherbener** [comment]
I am testing on Discover too. concretize gets a failure:
```
sherbene@discover12:/discover/nobackup/sherbene/projects/spack-stack/envs/unified-dev.discover-scu16> spack concretize --force --fresh 2>&1 | tee log.concretize
==> Error: concretization failed for the following reasons:

   1. cannot satisfy a requirement for package 'flex'.
==> Fetching https://mirror.spack.io/bootstrap/github-actions/v0.5/build_cache/linux-centos7-x86_64-gcc-10.2.1-clingo-bootstrap-spack-bhqgwuvef354fwuxq7heeighavunpber.spec.json
==> Fetching https://mirror.spack.io/bootstrap/github-actions/v0.5/build_cache/linux-centos7-x86_64/gcc-10.2.1/clingo-bootstrap-spack/linux-centos7-x86_64-gcc-10.2.1-clingo-bootstrap-spack-bhqgwuvef354fwuxq7heeighavunpber.spack
==> Installing "clingo-bootstrap@=spack%gcc@=10.2.1~docs+ipo+optimized+python+static_libstdcpp build_system=cmake build_type=Release generator=make patches=bebb819,ec99431 arch=linux-centos7-x86_64" from a buildcache
```
I think the issue is that common/packages.yaml has:
```
    flex:
      # Pin version to avoid duplicates
      require: '@2.6.4'
```
while site/packages.yaml has:
```
  flex:
    # Must set buildable: false to avoid duplicate packages
    buildable: false
    externals:
    - spec: flex@2.5.37+lex
      prefix: /usr
```
For now I can make the flex version match in common/packages.yaml in my environment area. But does that mean that we need to change something in this PR?

**Comment by climbfuji** [comment]
> I am testing on Discover too. concretize gets a failure:
> 
> ```
> sherbene@discover12:/discover/nobackup/sherbene/projects/spack-stack/envs/unified-dev.discover-scu16> spack concretize --force --fresh 2>&1 | tee log.concretize
> ==> Error: concretization failed for the following reasons:
> 
>    1. cannot satisfy a requirement for package 'flex'.
> ==> Fetching https://mirror.spack.io/bootstrap/github-actions/v0.5/build_cache/linux-centos7-x86_64-gcc-10.2.1-clingo-bootstrap-spack-bhqgwuvef354fwuxq7heeighavunpber.spec.json
> ==> Fetching https://mirror.spack.io/bootstrap/github-actions/v0.5/build_cache/linux-centos7-x86_64/gcc-10.2.1/clingo-bootstrap-spack/linux-centos7-x86_64-gcc-10.2.1-clingo-bootstrap-spack-bhqgwuvef354fwuxq7heeighavunpber.spack
> ==> Installing "clingo-bootstrap@=spack%gcc@=10.2.1~docs+ipo+optimized+python+static_libstdcpp build_system=cmake build_type=Release generator=make patches=bebb819,ec99431 arch=linux-centos7-x86_64" from a buildcache
> ```
> 
> I think the issue is that common/packages.yaml has:
> 
> ```
>     flex:
>       # Pin version to avoid duplicates
>       require: '@2.6.4'
> ```
> 
> while site/packages.yaml has:
> 
> ```
>   flex:
>     # Must set buildable: false to avoid duplicate packages
>     buildable: false
>     externals:
>     - spec: flex@2.5.37+lex
>       prefix: /usr
> ```
> 
> For now I can make the flex version match in common/packages.yaml in my environment area. But does that mean that we need to change something in this PR?

You should submit a site config update as a follow-up PR, yes. I expect we need a few more of those.

**Comment by srherbener** [comment]
Turns out that `/usr/bin/flex` on Discover is already version 2.6.4, so I updated the site/packages.yaml file (to version 2.6.4) and concretize successfully finished. The install is running now on Discover (and the one on S4 is still running). 

Do we still want to handle this update in a subsequent PR, or change this one. Here's the change I made in site/packages.yaml:
```
flex:
    # Must set buildable: false to avoid duplicate packages
    buildable: false
    externals:
    - spec: flex@2.6.4          <---- updated version
      prefix: /usr
```

**Review by srherbener — APPROVED** [review]
The S4 and Discover builds are still running. They appear to be in good shape. Since the concretize worked on both platforms (and the Discover correction is extremely simple), I think this PR is good to go, so I'm happy to approve. Thank you for doing this work!

**Comment by climbfuji** [comment]
@AlexanderRichert-NOAA Would you mind changing the flex version in the discover site config, as noted by @srherbener, before merging?

**Comment by srherbener** [comment]
I noticed that discover-scu17 already has this:
```
flex:
    # Must set buildable: false to avoid duplicate packages
    buildable: false
    externals:
    - spec: flex@2.6.4+lex
      prefix: /usr
```
Does the scu16 config also need the "+lex" suffix? I had started a feature branch, but I think it would be simpler to make the change in this PR and be done with it. Thanks!

**Comment by climbfuji** [comment]
I don't know how it was built on scu16

**Comment by srherbener** [comment]
Well with my testing we know that it works without the suffix, so we should probably leave it that way. I'll make a PR out of my feature branch and if @AlexanderRichert-NOAA decides to update this one, we can close the one I make. Thanks!

**Comment by AlexanderRichert-NOAA** [comment]
I see flex@2.6.4 in both discover configs. Is there a further change that needs to get made?

**Comment by climbfuji** [comment]
No, I merged the update from Steve in the meanwile. I also updated your branch from develop and enabled automerge - just waiting for CI to run.

**Comment by AlexanderRichert-NOAA** [comment]
:cowboy_hat_face:

**Pull request merged**
